Remove field refs from DatasetEditor

ranquild opened the pull request:
Depends on https://github.com/metabase/metabase/pull/45109
Related https://github.com/metabase/metabase/issues/42203
Related https://github.com/metabase/metabase/issues/25885
Related https://github.com/metabase/metabase/issues/33959
Related https://github.com/metabase/metabase/issues/36364

An attempt to simplify things by making metadata editing unaware of field refs. If this gets merged, we would rely on unique `column.name`. The PR doesn't fix anything on its own now.

Basically:
- `metadataDiff` becomes the single source of truth
- we use `column.name` for keys in `metadataDiff` and not field refs (which are not unique)
- we no longer change `result_metadata` in the current card before actually saving it
- we no longer try to make `metadataDiff` in sync with `result_metadata`. Instead, `metadataDiff` is applied on top of `result_metadata` in a selector. This was already the case, but because previously this code updated `result_metadata` at the same time, it had to deal with sync issues.

bshepherdson committed:
[QP] Include column aliases on `data.cols` and `result_metadata`

bshepherdson committed:
fix BE tests

ranquild committed:
Remove field_ref from DatasetEditor

ranquild committed:
Remove field_ref from DatasetEditor

ranquild committed:
Fix diff

ranquild commented on frontend/src/metabase-lib/v1/metadata/utils/models.ts:
```diff
@@ -148,42 +148,21 @@ export function getModelCacheSchemaName(databaseId: number, siteUUID: string) {
-    }))
-    .sort((a, b) => a.index - b.index)
-    .map(({ column }) => column);
+  return findColumnIndexesForColumnSettings(columns, columnSettings)
```
Rely on the function used for viz settings in the main part of the product; it's much more sophisticated

ranquild commented on frontend/src/metabase/query_builder/components/DatasetEditor/DatasetEditor.jsx:
```diff
@@ -93,6 +94,7 @@ const TABLE_HEADER_HEIGHT = 45;
 function mapStateToProps(state) {
   return {
     metadata: getMetadata(state),
+    visualizationSettings: getVisualizationSettings(state),
```
Rely on the computed viz settings. This handles missing columns and backfilling `table.columns` when needed.

ranquild committed:
Remove not needed wrong code

kamilmielnik commented on frontend/src/metabase-lib/v1/metadata/utils/models.ts:
```diff
@@ -3,14 +3,14 @@ import type Question from "metabase-lib/v1/Question";
 import { getQuestionVirtualTableId } from "metabase-lib/v1/metadata/utils/saved-questions";
 import type NativeQuery from "metabase-lib/v1/queries/NativeQuery";
-import { isSameField } from "metabase-lib/v1/queries/utils/field-ref";
+import { findColumnIndexesForColumnSettings } from "metabase-lib/v1/queries/utils/dataset";
```
> Make sure you can change metadata for duplicated columns separately

It does not seem to work. It's broken in `master` as well but has slightly different symptoms.
Are you sure this PR can fix it?

https://github.com/metabase/metabase/assets/6830683/f1420a5a-9356-49ad-a151-cf9d96ca17b1

ranquild replied:
@kamilmielnik you're correct - this PR is blocked until `column.name` is fixed and actually unique. Discussed here https://metaboat.slack.com/archives/C04CYTEL9N2/p1720016104124849

ranquild replied:
Fixed now

ranquild replied:
Not fixed. I reverted this change as it becomes too much for one PR

bshepherdson committed:
[QP] Disambiguate names in `:results_metadata`, eg. `ID_2`

This makes the names inconsistent with what's in `data.cols` on the same
API response(!) but also with what MBQL lib will return.

ranquild committed:
Merge remote-tracking branch 'origin/qp-disambiguate-names-in-result-metadata' into metadata-editor-cleanup

ranquild committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into metadata-editor-cleanup

ranquild committed:
Add repro

ranquild committed:
Try with column names

ranquild committed:
Fix column & column setting matching

ranquild committed:
Fix column & column setting matching

ranquild committed:
Revert broken changes

ranquild committed:
First attempt

ranquild committed:
First attempt

ranquild committed:
Another attempt

ranquild committed:
Another attempt

ranquild committed:
Another attempt

ranquild committed:
Another attempt

ranquild committed:
Another attempt

ranquild committed:
Add comments

ranquild committed:
Add comments

ranquild committed:
Add comments

ranquild committed:
Remove column setting keys

ranquild committed:
Add unit tests

ranquild commented on e2e/test/scenarios/models/reproductions.cy.spec.ts:
```diff
@@ -668,3 +668,90 @@ describe("issue 39749", () => {
+    cy.findByLabelText("Display name").should("have.value", name);
+  }
+
+  it("should allow to edit metadata for mbql models with self joins columns (metabase#25885)", () => {
```
Just to make it explicit, I'm not closing the issue as metadata reconciliation is broken on the BE right now.

ranquild committed:
Merge aliases

ranquild committed:
Revert

ranquild committed:
Revert

ranquild committed:
Revert

ranquild committed:
Revert "[QP] Include column aliases on `data.cols` and `result_metadata`"

This reverts commit 04b9dd59a7545a6d6a45e1b8ba2cdbf56bce5af7.

ranquild committed:
Revert

ranquild committed:
Revert

ranquild committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into metadata-editor-cleanup

ranquild committed:
Revert

ranquild committed:
Revert

ranquild committed:
Revert

ranquild committed:
Revert

ranquild committed:
Revert

ranquild committed:
Fix types

kamilmielnik commented on src/metabase/query_processor/middleware/results_metadata.clj:
```diff
@@ -77,6 +78,8 @@
              (map? result)
              (update :data
                      assoc
+                     ;; TODO: We agreed on the name `:result_metadata` everywhere, and this needs updating.
+                     ;; It'll definitely break things on the FE, so a coordinated change is needed.
```
Let's make sure there's a GH issue for that
